nncf/experimental/torch/fracbits/quantizer.py

```diff
@@ -0,0 +1,216 @@
+"""
+ Copyright (c) 2022 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+# Reference: Yang, Linjie, and Qing Jin. "Fracbits: Mixed precision quantization via fractional bit-widths."
+# Proceedings of the AAAI Conference on Artificial Intelligence. Vol. 35. No. 12. 2021.
+
+from typing import Dict
+import torch
+
+from nncf.experimental.torch.fracbits.structs import FracBitsQuantizationMode
+
+from nncf.torch.layer_utils import COMPRESSION_MODULES, CompressionParameter
+from nncf.torch.quantization.layers import (
+    QUANTIZATION_MODULES, AsymmetricQuantizer, PTQuantizerSpec, SymmetricQuantizer)
+from nncf.torch.quantization.quantize_functions import asymmetric_quantize, symmetric_quantize
+from nncf.torch.utils import no_jit_trace
+
+
+@COMPRESSION_MODULES.register()
+@QUANTIZATION_MODULES.register(FracBitsQuantizationMode.SYMMETRIC)
+class FracBitsSymmetricQuantizer(SymmetricQuantizer):
+    def __init__(self, qspec: PTQuantizerSpec):
+        super().__init__(qspec)
+        self._min_num_bits = int(0.5 * qspec.num_bits)
+        self._max_num_bits = int(1.5 * qspec.num_bits)
+        self._num_bits = CompressionParameter(torch.FloatTensor([qspec.num_bits]), requires_grad=True,
+                                              compression_lr_multiplier=qspec.compression_lr_multiplier)
+
+    @property
+    def frac_num_bits(self):
+        return torch.clamp(self._num_bits, self._min_num_bits, self._max_num_bits)
+
+    @property
+    def num_bits(self):
+        if self._num_bits.dtype == torch.int32:
```

Comment:
Could `torch.IntTensor` be always `torch.int32`? This can be `torch.qint32` from https://pytorch.org/docs/stable/tensors.html, can't this?

Reply:
It is constructed with `torch.int32` and it will not change unless explicitly changed to `torch.qint32`.

https://github.com/openvinotoolkit/nncf/blob/7827f19d23e64bf34b41a38c5d631a815a7df7f2/nncf/torch/quantization/layers.py#L272-L273